staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,988 @@
+	// scoring allows users to define a preference, without making it a
+	// hard requirement.
+	//
+	// Score *SomeScoringStruct
+}
+
+// Request is a request for one of many resources required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for several identical devices. It might get extended to support
+// asking for one of several different alternatives.
+type Request struct {
+	// The name can be used to reference this request in a pod.spec.containers[].resources.claims
+	// entry and in a constraint of the claim.
+	//
+	// Must be a DNS label.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	*RequestDetail `json:",inline" protobuf:"bytes,2,name=requestDetail"`
+
+	// FUTURE EXTENSION:
+	//
+	// OneOf contains a list of requests, only one of which must be satisfied.
+	// Requests are listed in order of priority.
+	//
+	// +optional
+	// +listType=atomic
+	// OneOf []RequestDetail
+}
+
+// RequestDetail is embedded inside Request. Exactly one field must be set.
+type RequestDetail struct {
+	// Device requests one or more devices.
+	//
+	// +required
+	Device *DeviceRequest `json:"device,omitempty" protobuf:"bytes,1,opt,name=device"`
+}
+
+// DeviceRequest is currently the only permitted alternative in RequestDetail.
+type DeviceRequest struct {
+	// By referencing a DeviceClass, a request inherits additional
+	// configuration parameters and selectors.
+	//
+	// A class is required. Which classes are available depends on the cluster.
+	//
+	// Administrators may use this to restrict which devices may get
+	// requested by only installing classes with selectors for permitted
+	// devices. If users are free to request anything without restrictions,
+	// then an empty class called "none" can get created to permit
+	// `deviceClassName: none`.
+	//
+	// +required
+	DeviceClassName string `json:"deviceClassName" protobuf:"bytes,1,name=deviceClassName"`
+
+	// Each selector must be satisfied by a device which is requested.
+	//
+	// +optional
+	// +listType=atomic
+	Selectors []Selector `json:"selectors,omitempty" protobuf:"bytes,2,name=selectors"`
+
+	// Amount determines how many devices to allocate for the request.
+	// The default if unset is exactly one device.
+	//
+	// +optional
+	Amount *Amount `json:"amount,omitempty" protobuf:"bytes,3,opt,name=amount"`
+
+	// AdminAccess indicates that this is a claim for administrative access
+	// to the device(s). Claims with AdminAccess are expected to be used for
+	// monitoring or other management services for a device.  They ignore
+	// all ordinary claims to the device with respect to access modes and
+	// any resource allocations. Ability to request this kind of access is
+	// controlled via ResourceQuota in the resource.k8s.io API.
+	//
+	// Default is false.
+	//
+	// +optional
+	AdminAccess *bool `json:"adminAccess,omitempty" protobuf:"bytes,4,opt,name=adminAccess"`
```

Comment:
@thockin: elsewhere you said "pointer for optional values", which is why I am doing that here.

But that then leads to three different cases:
- nil (implies false)
- pointer to false
- pointer to true

I ended up always accessing this field as `ptr.Deref(request.AdminAccess, false)`, which means that the client needs to know what the documented default for nil (`false` in this case) is.

We could add server-side defaulting so that `adminAccess: false` is always returned to clients. But IMHO that just makes `kubectl -o yaml` larger and isn't needed.

Wouldn't it be simpler to declare this as `AdminAccess bool`?

The same applies to strings.

I'm aware that the pointer allows us distinguish between "not set" and "set explicitly". I just don't see how knowing that is useful.

Reply:
>  >  I ended up always accessing this field as ptr.Deref(request.AdminAccess, false), which means that the client needs to know what the documented default for nil (false in this case) is.
>
> Don't do this, it's why we have defaults.

Unchecked pointer dereference always makes me nervous. That must be the C programmer in me. Even if there is a guarantee that it's never nil I feel compelled to check... :sweat_smile: 

> Especially for primitives, you can just say // +default=false (even for a plain bool) and then code will be generated if needed.

So isn't that then something that *has* to be added if the client shall not use `ptr.Deref(request.AdminAccess, false)`? As it stands right now, `request.AdminAccess` can be nil, so clients have to be prepared to deal with that, whether it is with `ptr.Deref` or an `if` check.

> Your call. ...  But this is not an existing API. Is this field something we want to elide in YAML when false?

Yes, because there's really no difference between "not set" and "explicitly set to false". If we keep it as a pointer and add defaulting to avoid encoding the default in the client, then all users will see an `adminAccess: false` in their `kubectl get - o yaml` although most don't use the feature.

So if I really have the choice here, then I prefer `AdminAccess bool`. The only reason that I see for keeping the pointers is that the protobuf encoding becomes more compact when the optional fields are unset (as noted on Slack). But both bool and empty string are short, so I'm not sure how much that matters.

> It's not like a string field which might have no value. We don't have consistent principles for that.

So let's continue with `ControllerName *string`. This cannot be the empty string (not valid), so I would prefer to also remove the pointer here. Okay?

Reply:
@thockin: GitHub "helpfully" keeps hiding this comment thread about pointers for basic types :sad:

I'm going ahead with removing the pointers. I can back it out again if there are still concerns about that.